### ISSUE TITLE
Handling BatchFixAllProvider crashes

### ISSUE THREAD

**Issue by paul1956** [dotnet/roslyn#7750]
I have a CodeFixProvider that is trying to fix all the instances of an issue in a Project (1,200 in my case). Memory keeps going up and I assume Visual Studio eventually runs our of memory and the BatchFixProvider gets an exception, then memory usage returns to normal. Is there a way that I should be detecting the user is trying to fix too many things at once and stop so that at least some are fixed. If I fix all the issues in a document, and manually process all the documents in a the project everything works so I don't think it is something I am doing.

![image](https://cloud.githubusercontent.com/assets/12789369/12069755/b902ac3e-affe-11e5-841a-1105d6cdecec.png)

**Comment by davkean** [comment]
@jmarolf @srivatsn Do you have suggestions for Paul?

**Comment by srivatsn** [comment]
We'll have to take a look at the BatchFixAllProvider and see why it's consuming so much memory. The only workaround that I can think of that wouldn't require a new roslyn is that you could implement your own FixAllProvider if there's an easy way to fix all instances for your fixer. Or I know @sharwell wrote a FixAllProvider that fixes all issues in a document in shot and repeats that. Since you say that works for you, you could try using his FixAllProvider.

**Comment by sharwell** [comment]
The problem lies in the way the batch fixer determines which diagnostics to fix. We unintentionally fixed it in DotNetAnalyzers/StyleCopAnalyzers@39948d9d5d574a4bd00a393ffafb3c3eed571787 while trying to do something else.

@paul1956 I recommend you take our [`CustomBatchFixAllProvider`](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/CustomBatchFixAllProvider.cs) (which is just `BatchFixAllProvider` with some bug fixes), [`CustomFixAllProviders`](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/CustomFixAllProviders.cs) and [`FixAllContextHelper`](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/FixAllContextHelper.cs) (a dependency of our batch fixer). Then [update your test suite](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/544b14b41df268c787dc86e7ae67a251ab693550/StyleCop.Analyzers/StyleCop.Analyzers.Test/Verifiers/CodeFixVerifier.cs#L88) to ensure you aren't shipping anything that depends on the buggy implementation.

**Comment by srivatsn** [comment]
@sharwell - it would be nice if you could send us some PRs for those bug fixes to the BatchFixAllProvider

**Comment by sharwell** [comment]
@srivatsn Many of the "bug fixes" were differences in the way we handle generated code files (#2753, #4390) and testing (#4416). The former should be resolved as part of the ongoing work to move generated code configuration into the hands of users. The primary performance issue actually lies in `FixAllContext.GetAllDiagnosticsAsync` and not the batch fixer, so it should be fixed in 1.2 without changes to the fixer itself (#7720).

**Comment by paul1956** [comment]
As @Sharwell said calling GetAllDiagnosticAsyn for every fix to make sure the fix did not cause a regression or fix multiple issues is extremely slow on a large solution minutes per loop but you never run out of memory so I am using that as a workaround. I will look at using your fixes to hopefully get the speed up also. But my issue was not about performance it was about memory consumption and crashing.

**Comment by paul1956** [comment]
I tried porting @sharwell to version of CustomBatchFixAllProvider to VB
```
    Public Overridable Overloads Async Function GetFixAsync(ByVal documentsAndDiagnosticsToFixMap As ImmutableDictionary(Of Document, ImmutableArray(Of Diagnostic)), ByVal fixAllContext As FixAllContext) As Task(Of CodeAction)
        If documentsAndDiagnosticsToFixMap IsNot Nothing AndAlso documentsAndDiagnosticsToFixMap.Any() Then
            fixAllContext.CancellationToken.ThrowIfCancellationRequested()

            Dim documents As ImmutableArray(Of Document) = documentsAndDiagnosticsToFixMap.Keys.ToImmutableArray()
            Dim options As ParallelOptions = New ParallelOptions() With {.CancellationToken = fixAllContext.CancellationToken}
            Dim fixesBag As List(Of CodeAction)() = New List(Of CodeAction)(documents.Length - 1) {}
            Parallel.ForEach(documents, options, Sub(document As Document, state As ParallelLoopState, index As Integer)
                                                     fixAllContext.CancellationToken.ThrowIfCancellationRequested()
                                                     fixesBag(index) = New List(Of CodeAction)()
                                                     Me.AddDocumentFixesAsync(document, documentsAndDiagnosticsToFixMap(document), fixesBag(index).Add, fixAllContext).Wait(fixAllContext.CancellationToken)
                                                 End Sub)

            If fixesBag.Any(Function(fixes) fixes.Count > 0) Then
                Return Await Me.TryGetMergedFixAsync(fixesBag.SelectMany(Function(i) i), fixAllContext).ConfigureAwait(False)
            End If
        End If

        Return Nothing
    End Function
```
But for both of the Parallel.ForEach I get an error and it is not obvious on what the issue is or how I Specify the data type(s) explicitly when I think they are.
```
Severity	Code	Description
Error 	BC30518	Overload resolution failed because no accessible 'ForEach' can be called with these arguments:
    'Public Shared Overloads Function ForEach(Of TSource)(source As Partitioner(Of TSource), parallelOptions As ParallelOptions, body As Action(Of TSource)) As ParallelLoopResult': Data type(s) of the type parameter(s) cannot be inferred from these arguments. Specifying the data type(s) explicitly might correct this error.
    'Public Shared Overloads Function ForEach(Of TSource)(source As Partitioner(Of TSource), parallelOptions As ParallelOptions, body As Action(Of TSource, ParallelLoopState)) As ParallelLoopResult': Data type(s) of the type parameter(s) cannot be inferred from these arguments. Specifying the data type(s) explicitly might correct this error.
    'Public Shared Overloads Function ForEach(Of TSource)(source As OrderablePartitioner(Of TSource), parallelOptions As ParallelOptions, body As Action(Of TSource, ParallelLoopState, Long)) As ParallelLoopResult': Data type(s) of the type parameter(s) cannot be inferred from these arguments. Specifying the data type(s) explicitly might correct this error.
```

**Comment by paul1956** [comment]
@sharwell your fix is much faster and uses less memory, I have seen no issues with Document, the one Project I tried consumes over 3 gig (compared to Out of Memory Error) but solution still runs out of memory.

**Comment by paul1956** [comment]
@davkean this should be a bug not a question and is unrelated to the issues @sharwell raised, though his fixes improved the issue.

**Comment by jmarolf** [comment]
@paul1956 could you share your codefix with us?  It sounds like you are creating 2-3MB of objects per diagnostic.  If we knew which apis you were using we could look into ways to reduce memoty pressure for those apis.

**Comment by paul1956** [comment]
@jmarolf I have attached a VB project that shows the OOM issue. Ignore the ToolWindow it was a poor workaround for OOM issue. The 3 files of interest are in the SummaryWindow\SummaryWindow\Refactoring directory everything else is stuff to make the VSIX and test.
[SummaryWindow.zip](https://github.com/dotnet/roslyn/files/97103/SummaryWindow.zip)

**Comment by jmarolf** [comment]
@paul1956 thanks, could you include your Extensions folder in the zip?

**Comment by paul1956** [comment]
@jmarolf how about just stopping applying additional fixes when you run out of memory (or are about too). Of allow me to specify how many fixes to apply at one time, project is too many and document is too few. I have noticed that even when I do Document it does not get them all on one pass anyway.

**Comment by paul1956** [comment]
@jmarolf the only extension needed is in there. The only stuff I deleted was binaries and packages because they were to large to upload in 10 megabytes.

**Comment by paul1956** [comment]
I ran a different test and noticed that during the mix memory is constant but when it tried to do the merge with either Roslyn code or @sharwell CustomBatchFixAllProvider memory increase rapidly. The flat area below is my code fixing the issues and speed raise at the end in the merge.
![image](https://cloud.githubusercontent.com/assets/12789369/12466277/3b119760-bf88-11e5-95e2-67d042a7fb14.png)

**Comment by paul1956** [comment]
@jmarolf memory is running out in the Roslyn function (TryMergeFixesAsync) below when I enter the function memory used is 1.5 Gig. 
```
Memory Usage (Working Set): 1,503,664K VM, Size (Private Bytes): 1,350,536, K GC Total Memory: 581,937,444 bytes
```
After 134 CodeAction loops
```
Memory Usage (Working Set): 1,904,064K VM, Size (Private Bytes): 1,753,124, K GC Total Memory: 1,398,892,516 bytes
```
This memory increase will continue until all the CodeActions are complete or you run out of memory.
```
public virtual async Task<Solution> TryMergeFixesAsync(Solution oldSolution, IEnumerable<CodeAction> codeActions, CancellationToken cancellationToken)
```
It seems that this function could monitor memory and exit gracefully before OOM and still fix a reasonable number of issues. Or it needs to be redesigned to use less memory.

**Comment by jmarolf** [comment]
@paul1956 before I can take a look, I can't build the project you uploaded. The following extension methods are not included.  
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\AnalizerExtensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\Extensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\GeneratedCodeAnalysisExtensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\KindExtensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\StringExtensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\VBAnalyzerExtensions.vb
..\..\VBCodeRefactoring\VBCodeRefactoring\VBCodeRefactoring\Extensions\VBGeneratedCodeAnalysisExtensions.vb

**Comment by paul1956** [comment]
@jmarolf Sorry did not know those files were links
[Extensions.zip](https://github.com/dotnet/roslyn/files/102860/Extensions.zip)

**Comment by paul1956** [comment]
@jmarolf were you able to build my sample and see the issue? I use  @sharwell customBatchFixProvider because it lets me debug and work around the issue but the issue happens with the standard BatchFixProvider. Any CodeFixProvider working on a large project or solution with lots of changes will hit this.

**Comment by CyrusNajmabadi** [comment]
This has been rewritten and should be fine here.  We also recommend usingFixAllProvider.Create to get an even more efficient fixer that can work in parallel across documents.
